homeassistant/components/icloud.py

```diff
@@ -0,0 +1,946 @@
+
+    if not ICLOUDTRACKERS:
+        _LOGGER.error("No ICLOUDTRACKERS added")
+        return False
+
+    randomseconds = random.randint(10, 59)
+
+    def lost_iphone(call):
+        """Call the lost iphone function if the device is found."""
+        accountname = call.data.get('accountname')
+        devicename = call.data.get('devicename')
+        if accountname is None:
+            for account in ICLOUDTRACKERS:
+                ICLOUDTRACKERS[account].lost_iphone(devicename)
+        elif accountname in ICLOUDTRACKERS:
+            ICLOUDTRACKERS[accountname].lost_iphone(devicename)
+
+    hass.services.register(DOMAIN, 'lost_iphone',
+                           lost_iphone)
+
+    def update_icloud(call):
+        """Call the update function of an icloud account."""
+        accountname = call.data.get('accountname')
+        devicename = call.data.get('devicename')
+        if accountname is None:
+            for account in ICLOUDTRACKERS:
+                ICLOUDTRACKERS[account].update_icloud(devicename)
+        elif accountname in ICLOUDTRACKERS:
+            ICLOUDTRACKERS[accountname].update_icloud(devicename)
+    hass.services.register(DOMAIN,
+                           'update_icloud', update_icloud)
+
+    def keep_alive(now):
+        """Keep the api logged in of all account."""
+        for accountname in ICLOUDTRACKERS:
+            try:
+                ICLOUDTRACKERS[accountname].keep_alive()
+            except ValueError:
+                _LOGGER.error("something went wrong for account %s, " +
+                              "retrying in a minute", accountname)
+
+    track_utc_time_change(
+        hass, keep_alive,
+        second=randomseconds
+    )
+
+    def setinterval(call):
+        """Call the update function of an icloud account."""
+        accountname = call.data.get('accountname')
+        interval = call.data.get('interval')
+        devicename = call.data.get('devicename')
+        if accountname is None:
+            for account in ICLOUDTRACKERS:
+                ICLOUDTRACKERS[account].setinterval(interval, devicename)
+        elif accountname in ICLOUDTRACKERS:
+            ICLOUDTRACKERS[accountname].setinterval(interval, devicename)
+
+    hass.services.register(DOMAIN,
+                           'setinterval', setinterval)
+
+    # Tells the bootstrapper that the component was successfully initialized
+    return True
+
+
+class IDevice(Entity):  # pylint: disable=too-many-instance-attributes
+    """Represent an iDevice in Home Assistant."""
+
+    def __init__(self, hass, icloudobject, name, identifier, googletraveltime):
+        """Initialize the iDevice."""
+        # pylint: disable=too-many-arguments
+        self.hass = hass
+        self.icloudobject = icloudobject
+        self.devicename = name
+        self.identifier = identifier
+        self._request_interval_seconds = 60
+        self._interval = 1
+        self.api = icloudobject.api
+        self._distance = None
```

Comment:
You build it later on where you receive the values...

Here you only create the dict instead of all these `=None` assignments
```python
self._attrs = {}
```

You populate it as you interrogate iCloud on other places...
```
self._attrs[ATTR_DEVICENAME] = the_value
# and for almost all of your ATTR_s
```

and then the complete `state_attributes` function becomes:
```python
def state_attributes(self):
    """Return the friendlyname of the iDevice."""
    return self._attrs
```